[libpas] Extend Adjacent-Tag-Exclusion to cover bitfit objects as well

Achierius opened the pull request:
#### e698dda8251bff93d56d3df5601040aacbc36155
<pre>
[libpas] Extend Adjacent-Tag-Exclusion to cover bitfit objects as well
<a href="https://bugs.webkit.org/show_bug.cgi?id=302810">https://bugs.webkit.org/show_bug.cgi?id=302810</a>
<a href="https://rdar.apple.com/164245530">rdar://164245530</a>

Reviewed by Dan Hecht.

Due to how a single bitfit page can contain objects of varying size,
ensuring adjacent-tag-exclusion comes with a few extra challenges.
 1. We cannot use the parity trick to divide allocations into
    odd/even, so instead we directly ldg the previous/subsequent
    allocations and filter those tags out of the irg computation.
 2. We cannot ldg across pages, so we need to be a bit clever to ensure
    that adjacent allocations on different pages do not collide. We do
    so by having all bitfit page allocations be even-tagged, save for
    the allocation (if any) containing the last MTE granule in the page,
    which is odd-tagged.
Together, this ensures that the ATE property should apply to bitfit
objects both within and across allocator pages.

Canonical link: <a href="https://commits.webkit.org/304033@main">https://commits.webkit.org/304033@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/215b4525226ede528491cd16cc9902b73bd65246

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134382 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/6848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45595 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/141917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7f62f9de-c2eb-40e0-ae82-bca6018f82f3) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7445 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/6715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/141917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/49740bf2-7408-477c-95fb-9bdf7bcac17a) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137329 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/5193 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/120432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/141917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f826f0f5-3753-4d35-8984-b7746c32a818) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/5055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [⏳ 🛠 wpe-cairo-libwebrtc ](https://ews-build.webkit.org/#/builders/WPE-Cairo-LibWebRTC-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126438 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/38573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144606 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132890 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/6528 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/39156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/6604 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/6715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111389 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28247 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/116708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/60310 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/6580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165821 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/6405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/6640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/6517 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -65,9 +66,10 @@ PAS_IGNORE_WARNINGS_BEGIN("unsafe-buffer-usage")
  * from OpenSource's pas_internal_config.h -- we cannot use it directly as
  * pas_utils.h is too high up in the include hierarchy.
  */
-#define PAS_MTE_SMALL_PAGE_DEFAULT_SHIFT (14)
```
Previously we duplicated this here because we couldn't include pas_internal_config.h, but now we can so we can drop that

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -195,6 +193,95 @@ enum pas_mte_tag_constraint {
   pas_mte_nonzero_even_tag = 0xaaab,
```
I think a quick comment explaining what pas_mte_tag_constraint is would be helpful. It wasn't immediately obvious to me until reading more code what it is. Maybe:

Bitset of tags in the range 0-15 indicating the disallowed set of tags for each constraint.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -195,6 +193,95 @@ enum pas_mte_tag_constraint {
+
+        // Rely on objects being MTE-granule-aligned and on
+        // sizes being a multiple of the granule size
+        uintptr_t prior_granule = ptr - 16;
```
Is there a define for granule size we can use instead of 16?

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -195,6 +193,95 @@ enum pas_mte_tag_constraint {
+        // tag generation.
+
+        // Rely on objects being MTE-granule-aligned and on
+        // sizes being a multiple of the granule size
```
maybe PAS_TEST_ASSERT those assumptions?

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -195,6 +193,95 @@ enum pas_mte_tag_constraint {
+            PAS_MTE_GET_MTAG(prior_granule);
+            valid_tags = pas_mte_exclude_tag(valid_tags, prior_granule >> PAS_MTE_TAG_SHIFT);
+        }
+    }
```
I found it a bit hard to reason about the possible combinations with this disjoint logic, especially with the negation of the above. Consider being explicit about the three cases, something like:

```
  // Determine which adjacent tag to exclude based on page boundaries
  if (current_pageno != prior_pageno) {
      // Allocation starts at beginning of page
      PAS_ASSERT(current_pageno == succeeding_pageno);
      valid_tags = pas_mte_nonzero_even_tag;
      PAS_MTE_GET_MTAG(succeeding_granule);
      valid_tags = pas_mte_exclude_tag(valid_tags, succeeding_granule >> PAS_MTE_TAG_SHIFT);
  } else if (current_pageno != succeeding_pageno) {
      // Allocation ends at end of page
      valid_tags = pas_mte_odd_tag;
      PAS_MTE_GET_MTAG(prior_granule);
      valid_tags = pas_mte_exclude_tag(valid_tags, prior_granule >> PAS_MTE_TAG_SHIFT);
  } else {
      // Allocation is in middle of page (both neighbors on same page)
      valid_tags = pas_mte_any_nonzero_tag;
      PAS_MTE_GET_MTAG(succeeding_granule);
      valid_tags = pas_mte_exclude_tag(valid_tags, succeeding_granule >> PAS_MTE_TAG_SHIFT);
  }
```

dhecht replied:
If you don't rewrite it that way, you should at least fix the line break between 276/277

dhecht replied:
Oh, or maybe that `else` on line 276 is actually a bug? We should exclude both in that case, right? if so, I think just removing that `else` would make this code make sense as-is.

Achierius replied:
Fixed

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -195,6 +193,95 @@ enum pas_mte_tag_constraint {
+    uintptr_t ptr,
+    size_t size,
+    pas_allocator_homogeneity homogeneity,
+    bool is_known_medium)
```
I don't follow what the "known" part of this name means. Does it mean that if this is false, it might be small or medium, we just didn't know it? How does the logic of this function work if that's the case?

Achierius replied:
At one point it did, but that is no longer the case AFAICT -- renaming it was part of the refactoring that I took out of this patch to save for later.

Achierius committed:
[libpas] Extend Adjacent-Tag-Exclusion to cover bitfit objects as well
https://bugs.webkit.org/show_bug.cgi?id=302810
rdar://164245530

Reviewed by Dan Hecht.

Due to how a single bitfit page can contain objects of varying size,
ensuring adjacent-tag-exclusion comes with a few extra challenges.
 1. We cannot use the parity trick to divide allocations into
    odd/even, so instead we directly ldg the previous/subsequent
    allocations and filter those tags out of the irg computation.
 2. We cannot ldg across pages, so we need to be a bit clever to ensure
    that adjacent allocations on different pages do not collide. We do
    so by having all bitfit page allocations be even-tagged, save for
    the allocation (if any) containing the last MTE granule in the page,
    which is odd-tagged.
Together, this ensures that the ATE property should apply to bitfit
objects both within and across allocator pages.

Canonical link: https://commits.webkit.org/304033@main
